File: Oc-Dome/Oc-Dome/Classes/Main/BaseViewController.m
//
//  BaseViewController.m
//  kzyjsq
//
//  Created by 李含文 on 2018/12/1.
//  Copyright © 2018年 李含文. All rights reserved.
//

#import "BaseViewController.h"

@interface BaseViewController ()<UIGestureRecognizerDelegate>

@end

@implementation BaseViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    [self setBGColor];
}

- (void)viewWillAppear:(BOOL)animated{
    [super viewWillAppear:animated];
    HWLog(@"进入%@",[self class]);
    [self switchWhiteColor];
}

- (void)viewDidDisappear:(BOOL)animated {
    [super viewDidDisappear:animated];
    //    [HUDManage dismiss];
}

- (void)dealloc {
    HWLog(@"释放%@",[self class]);
}
// MARK: - 修改导航栏颜色
- (void)switchNavColor:(UIColor *)color {
    [self.navigationController.navigationBar setBackgroundImage:[UIImage hw_imageWithColor:color] forBarMetrics:UIBarMetricsDefault];
    if (color != [UIColor whiteColor]) {
        self.navigationController.navigationBar.shadowImage = [[UIImage alloc] init];
    }
    [UIApplication sharedApplication].statusBarStyle = (color == [UIColor whiteColor] ? UIStatusBarStyleDefault : UIStatusBarStyleLightContent);
    NSMutableDictionary *dic = [[NSMutableDictionary alloc] init];
    //    dic[NSFontAttributeName] = [UIFont systemFontOfSize:18];
    UIColor *setColor = (color == [UIColor whiteColor] ? [UIColor blackColor] : [UIColor whiteColor]);
    if (color == [UIColor clearColor]) {
        [self.navigationController.navigationBar setTranslucent:YES];
    } else {
        [self.navigationController.navigationBar setTranslucent:NO];
    }
    dic[NSForegroundColorAttributeName] = setColor;
    self.navigationController.navigationBar.titleTextAttributes = dic;
    self.navigationController.navigationBar.tintColor = setColor;
}

- (void)switchGradientColor {
    [self.navigationController.navigationBar setBackgroundImage:[UIImage hw_imageWithColor:hw_GradientColor] forBarMetrics:UIBarMetricsDefault];
    self.navigationController.navigationBar.shadowImage = [[UIImage alloc] init];
    [UIApplication sharedApplication].statusBarStyle = UIStatusBarStyleLightContent;
    NSMutableDictionary *dic = [[NSMutableDictionary alloc] init];
    //    dic[NSFontAttributeName] = [UIFont systemFontOfSize:18];
    [self.navigationController.navigationBar setTranslucent:NO];
    UIColor *setColor = [UIColor whiteColor];
    dic[NSForegroundColorAttributeName] = setColor;
    self.navigationController.navigationBar.titleTextAttributes = dic;
    self.navigationController.navigationBar.tintColor = setColor;
}
- (void)switchWhiteColor {
    [self switchNavColor:[UIColor whiteColor]];
}
- (void)switchBlackColor {
    [self switchNavColor:[UIColor blackColor]];
}
- (void)switchClearColor {
    [self switchNavColor:[UIColor clearColor]];
}

// MARK: - 快速添加返回按钮 并让边缘手势生效
- (void)setBackButton:(void (^)(void))action {
    UIButton *btn = [[UIButton alloc] initWithFrame:CGRectMake(0, 0, 50, 44)];
    [btn setImage:[UIImage hw_imageNamed:@"返回"] forState:(UIControlStateNormal)];
    btn.actionBlock = ^{
        action();
    };
    // 让返回按钮内容继续向左边偏移10
    btn.contentEdgeInsets = UIEdgeInsetsMake(0, -15, 0, 0);
    btn.imageEdgeInsets = UIEdgeInsetsMake(0, 0, 0, 30);
    UIBarButtonItem *backItem = [[UIBarButtonItem alloc] initWithCustomView:btn];
    self.navigationItem.leftBarButtonItem = backItem;
    //    UIBarButtonItem * spaceItem = [[UIBarButtonItem alloc]initWithBarButtonSystemItem:UIBarButtonSystemItemFixedSpace target:nil action:nil];
    //    //将宽度设为负值
    //    spaceItem.width = -15;
    //    self.navigationItem.leftBarButtonItems = @[spaceItem,backItem];
    // 自定义返回按钮是边缘手势会失效  实现下面方法即可
    if (self.navigationController.viewControllers.count > 1) {
        self.navigationController.interactivePopGestureRecognizer.delegate = self;
    }
}

@end

File: Swift-Dome/LHWDome/Classes/Main(主要基类)/BaseViewController.swift
//
//  BaseViewController.swift
//  BBCShop
//
//  Created by 梁琪琛 on 2018/2/9.
//  Copyright © 2018年 SK丿希望. All rights reserved.
//

import UIKit


class BaseViewController: UIViewController {
    private var themeColor = UIColor.white
    override func viewDidLoad() {
        switchNavColor(HWMainColor())
        super.viewDidLoad()
//        setBGColor()
    }
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        print("进入了---\(type(of: self))")
    }
    deinit {
        print("\(type(of: self))释放")
        NotificationCenter.default.removeObserver(self)
    }
    public func setBackBtn(_ tintColor: UIColor = .black) {
        let leftBarButtonItem = UIBarButtonItem.init(image: UIImage(named: "返回白"), style: UIBarButtonItem.Style.done, target: self, action: #selector(backBtnClick))//
        leftBarButtonItem.tintColor = tintColor
        if #available(iOS 11.0, *){ // ios11 以上偏移
            leftBarButtonItem.imageInsets = UIEdgeInsets(top: 0, left: -11, bottom: 0, right: 0)
            navigationItem.leftBarButtonItem = leftBarButtonItem
        } else {
            let nagetiveSpacer = UIBarButtonItem.init(barButtonSystemItem: UIBarButtonItem.SystemItem.fixedSpace, target: nil, action: nil)
            nagetiveSpacer.width = -11//这个值可以根据自己需要自己调整
            navigationItem.leftBarButtonItems = [nagetiveSpacer, leftBarButtonItem]
        }
        if (navigationController?.viewControllers.count ?? 0) > 1 {
            navigationController?.interactivePopGestureRecognizer?.delegate = self as? UIGestureRecognizerDelegate // 重要 自定义返回按钮恢复返回手势
        }
    }
    @objc private func backBtnClick() {
        dismiss(animated: true, completion: nil)
        navigationController?.popViewController(animated: true)
    }
    
    /// 修改导航栏颜色
    ///
    /// - Parameter color: 需要修改的颜色
    func switchNavColor(_ color: UIColor) {
        themeColor = color
        navigationController?.navigationBar.setBackgroundImage(UIImage(color: color, size: UIScreen.main.bounds.size), for: UIBarMetrics.default)
        navigationController?.navigationBar.shadowImage = UIImage()
        //                navigationController?.navigationBar.isTranslucent = true // 会觉得颜色很浅，因为这是半透明状态 iOS11 以下出现内容上偏
        navigationController?.navigationBar.isTranslucent = false // 推荐关闭半透明状态 导航栏位置 相对布局以0开始
        UIApplication.shared.statusBarStyle = color != UIColor.white ? .lightContent : .default
        navigationController?.navigationBar.titleTextAttributes =  [NSAttributedString.Key.foregroundColor: color != UIColor.white ? UIColor.white : UIColor.black, NSAttributedString.Key.font: UIFont.systemFont(ofSize:18)]
        navigationController?.navigationBar.tintColor = color != UIColor.white ? UIColor.white : UIColor.black //item 字体颜色
        setNeedsStatusBarAppearanceUpdate()
    }
    override var preferredStatusBarStyle: UIStatusBarStyle {
        return themeColor != UIColor.white ? .lightContent : .default
    }
}

